# Project Title: Diwali Sales Analytics

## Project Summary

The **Diwali Sales Analytics** project is designed to help businesses understand customer behavior, product performance, and regional sales patterns during the festive season. Using Python and libraries such as **pandas, numpy, matplotlib, and seaborn**, this project performs detailed data analysis on Diwali sales data to generate actionable insights for marketing, inventory optimization, and sales planning.

### The analysis focuses on:

### 1. Customer Demographics & Behavior
- Understanding purchasing patterns based on **age, gender, marital status, and occupation**.
- Identifying **repeat and high-value customers** to improve targeting strategies.

### 2. Product & Revenue Analysis
- Determining **top-selling products** and **high-performing product categories**.
- Analyzing **revenue trends**, daily sales patterns, and the **impact of discounts** on sales.

### 3. Regional Sales Patterns
- Evaluating **state-wise** and **zone-wise** sales performance.
- Identifying **regional preferences** for products and categories to optimize stock and marketing.

### 4. Advanced Insights
- Segmenting customers for **targeted marketing campaigns**.
- Visualizing product performance using **heatmaps and trend charts**.
- Forecasting **future Diwali sales** using historical trends.


## Problem Statement

During the festive season, businesses witness a significant surge in customer purchases. Understanding **customer behavior**, **product preferences**, and **regional sales patterns** is crucial to maximize revenue and plan effective, targeted marketing strategies.

This project analyzes **Diwali sales data** to uncover:

- Customer demographics and purchasing behavior based on **age, gender, marital status, and occupation**.
- **High-performing product categories** and overall **revenue trends**.
- **Regional sales patterns** across different **states and zones**.

The goal of this analysis is to help businesses make **data-driven decisions** for marketing, inventory planning, and sales optimization during the Diwali festive season.


## Objective

The objective of this project is to provide **actionable insights** that support:

- Effective **marketing strategies** through customer segmentation and behavior analysis.
- **Inventory optimization** by identifying high-demand products and categories.
- Improved **sales planning** by understanding revenue trends and regional sales performance during the Diwali festive season.


## Outcome

This project provides businesses with valuable insights to:

- **Increase revenue** by targeting the right customer segments.
- **Optimize inventory management** by stocking high-demand products.
- **Plan region-specific marketing campaigns** based on customer and sales trends.

## Technologies Used

- **Python**  
- **Jupyter Notebook**  
- **pandas**  
- **numpy**  
- **matplotlib**  
- **seaborn**


# 🎉 Let’s Dive Into Diwali Sales Analytics!
### Uncovering Customer Behavior, Product Trends, and Regional Insights During the Festive Season


## 1. Know Your Data

**What it is:**  
Examining the dataset to understand its structure, size, column types, and basic statistics.

**Why we use it:**  
To get an overall idea of the data, identify potential issues like missing values or incorrect data types, and plan the cleaning and analysis steps effectively.


#### Import Libraries

**What it is:**  
Loading Python libraries like **pandas, numpy, matplotlib, seaborn** that provide tools for data manipulation, analysis, and visualization.

**Why we use it:**  
These libraries make it easy to **read, clean, explore, and visualize data efficiently**, which is essential for any data analysis project.


In [1]:
# Import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Optional: To display plots inline in Jupyter Notebook
%matplotlib inline

# Set seaborn style
sns.set(style="whitegrid")

#### Load Dataset

**What it is:**  
Reading the data file (CSV/Excel) into a Python environment as a DataFrame using **pandas**.

**Why we use it:**  
To bring the dataset into the workspace so we can **analyze, manipulate, and visualize** it programmatically.


In [2]:
# Read the CSV file
df = pd.read_csv(r"C:\Users\Prafull Wahatule\Desktop\Diwali Sales Analytics - Pyhton\Data\Diwali Sales Data.csv")

#### Take a Look at Data

**What it is:**  
Using `df.head()` to display the first few rows of the dataset.

**Why we use it:**  
To get a **quick preview** of the data, understand column names, and check if the data has been loaded correctly.


In [3]:
# Display the first few rows
df.head()

User_ID  Cust_name Product_ID Gender Age Group  Age  Marital_Status  \
0  1002903  Sanskriti  P00125942      F     26-35   28               0   
1  1000732     Kartik  P00110942      F     26-35   35               1   
2  1001990      Bindu  P00118542      F     26-35   35               1   
3  1001425     Sudevi  P00237842      M      0-17   16               0   
4  1000588       Joni  P00057942      M     26-35   28               1   

            State      Zone       Occupation Product_Category  Orders  \
0     Maharashtra   Western       Healthcare             Auto       1   
1  Andhra Pradesh  Southern             Govt             Auto       3   
2   Uttar Pradesh   Central       Automobile             Auto       3   
3       Karnataka  Southern     Construction             Auto       2   
4         Gujarat   Western  Food Processing             Auto       2   

    Amount  Status  unnamed1  
0  23952.0     NaN       NaN  
1  23934.0     NaN       NaN  
2  23924.0     NaN       NaN  
3  23912.0     NaN       NaN  
4  23877.0     NaN       NaN

#### Dataset Rows & Columns Count

**What it is:**  
Checking the number of rows and columns in the dataset using `df.shape`.

**Why we use it:**  
To understand the **size and dimensions** of the dataset, which helps in planning the analysis and knowing how much data we are dealing with.

In [4]:
# Dataset Rows & Columns count
df.index

RangeIndex(start=0, stop=11251, step=1)

In [5]:
# Getting Information of total Column & Rows
df.shape

(11251, 15)

In [6]:
df.columns

Index(['User_ID', 'Cust_name', 'Product_ID', 'Gender', 'Age Group', 'Age',
       'Marital_Status', 'State', 'Zone', 'Occupation', 'Product_Category',
       'Orders', 'Amount', 'Status', 'unnamed1'],
      dtype='object')

#### Getting Dataset Information

**What it is:**  
Using `df.info()` to get details about column names, data types, non-null counts, and memory usage.

**Why we use it:**  
To quickly **identify data types, missing values, and overall structure** of the dataset, which helps in planning data cleaning and analysis steps.


In [7]:
# Dataset Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11251 entries, 0 to 11250
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   User_ID           11251 non-null  int64  
 1   Cust_name         11251 non-null  object 
 2   Product_ID        11251 non-null  object 
 3   Gender            11251 non-null  object 
 4   Age Group         11251 non-null  object 
 5   Age               11251 non-null  int64  
 6   Marital_Status    11251 non-null  int64  
 7   State             11251 non-null  object 
 8   Zone              11251 non-null  object 
 9   Occupation        11251 non-null  object 
 10  Product_Category  11251 non-null  object 
 11  Orders            11251 non-null  int64  
 12  Amount            11239 non-null  float64
 13  Status            0 non-null      float64
 14  unnamed1          0 non-null      float64
dtypes: float64(3), int64(4), object(8)
memory usage: 1.3+ MB


#### Duplicate Values

**What it is:**  
Checking for duplicate rows in the dataset using `df.duplicated().value_counts()`.

**Why we use it:**  
To **identify and remove duplicate records**, ensuring that analysis and insights are accurate and not skewed by repeated data.


In [8]:
# Dataset Duplicate Value Count
duplicate_values = df.duplicated().value_counts()  
duplicate_values

False    11243
True         8
Name: count, dtype: int64

## 2. Data Cleaning & Data Wrangling

**What it is:**  
The process of **handling missing values, removing duplicates, correcting data types, and dropping unnecessary columns** to make the dataset consistent and analysis-ready.

**Why we use it:**  
Clean data ensures **accurate, reliable, and meaningful analysis**. Without cleaning, insights may be incorrect or misleading.


#### Drop Columns

**What it is:**  
Removing unnecessary or irrelevant columns from the dataset using `df.drop()`.

**Why we use it:**  
To **simplify the dataset**, reduce clutter, and focus only on the columns that are important for analysis.


In [9]:
# Drop columns with all null values
df.drop(columns=['Status', 'unnamed1'], inplace=True)

In [10]:
# Check Current Columns
df.columns

Index(['User_ID', 'Cust_name', 'Product_ID', 'Gender', 'Age Group', 'Age',
       'Marital_Status', 'State', 'Zone', 'Occupation', 'Product_Category',
       'Orders', 'Amount'],
      dtype='object')

#### Fill Missing Amount Values

**What it is:**  
Replacing missing values in the `Amount` column using methods like `fillna()` with median or zero.

**Why we use it:**  
To ensure that **sales calculations and analysis are accurate**, since missing values can skew totals, averages, and other insights.


In [11]:
# Fill missing Amount values with 0 or median
df['Amount'] = df['Amount'].fillna(df['Amount'].median())

In [12]:
# Missing Values/Null Values Count
df.isnull().sum()

User_ID             0
Cust_name           0
Product_ID          0
Gender              0
Age Group           0
Age                 0
Marital_Status      0
State               0
Zone                0
Occupation          0
Product_Category    0
Orders              0
Amount              0
dtype: int64

#### Rename Columns to Remove Spaces

**What it is:**  
Renaming columns (e.g., `'Age Group'` → `'Age_Group'`) using `df.rename()` to remove spaces.

**Why we use it:**  
To make column names **easier to reference in code**, avoid syntax errors, and improve readability during analysis.


In [13]:
# Rename columns to remove spaces for easier coding
df.rename(columns={'Age Group':'Age_Group'}, inplace=True)

In [14]:
# Check Current Columns
df.columns

Index(['User_ID', 'Cust_name', 'Product_ID', 'Gender', 'Age_Group', 'Age',
       'Marital_Status', 'State', 'Zone', 'Occupation', 'Product_Category',
       'Orders', 'Amount'],
      dtype='object')

#### Check & Drop Duplicate Records

**What it is:**  
Identifying duplicate rows using `df.duplicated()` and removing them with `df.drop_duplicates()`.

**Why we use it:**  
To ensure that each transaction is **unique**, preventing inflated counts or skewed analysis results caused by repeated data.


In [15]:
#Check Duplicate Records
df.duplicated().value_counts()

False    11243
True         8
Name: count, dtype: int64

In [16]:
# Drop Duplicate Records
df.drop_duplicates(inplace=True)

#### Verify Data After Cleaning

**What it is:**  
Checking the dataset after cleaning using `df.info()`, `df.shape`, and `df.head()` to ensure all issues are resolved.

**Why we use it:**  
To **confirm that missing values are handled, duplicates are removed, and columns are correctly formatted**, making the dataset fully ready for analysis.


In [17]:
# Verify Data After Cleaning

df.info()
df.shape

<class 'pandas.core.frame.DataFrame'>
Index: 11243 entries, 0 to 11250
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   User_ID           11243 non-null  int64  
 1   Cust_name         11243 non-null  object 
 2   Product_ID        11243 non-null  object 
 3   Gender            11243 non-null  object 
 4   Age_Group         11243 non-null  object 
 5   Age               11243 non-null  int64  
 6   Marital_Status    11243 non-null  int64  
 7   State             11243 non-null  object 
 8   Zone              11243 non-null  object 
 9   Occupation        11243 non-null  object 
 10  Product_Category  11243 non-null  object 
 11  Orders            11243 non-null  int64  
 12  Amount            11243 non-null  float64
dtypes: float64(1), int64(4), object(8)
memory usage: 1.2+ MB


(11243, 13)

### Final Dataset Preview

**What it is:**  
Displaying the first few rows of the cleaned dataset using `df.head()`.

**Why we use it:**  
To **visually inspect** the cleaned data, verify column names, data types, and sample values before starting detailed analysis.


In [18]:
df.head()

User_ID  Cust_name Product_ID Gender Age_Group  Age  Marital_Status  \
0  1002903  Sanskriti  P00125942      F     26-35   28               0   
1  1000732     Kartik  P00110942      F     26-35   35               1   
2  1001990      Bindu  P00118542      F     26-35   35               1   
3  1001425     Sudevi  P00237842      M      0-17   16               0   
4  1000588       Joni  P00057942      M     26-35   28               1   

            State      Zone       Occupation Product_Category  Orders   Amount  
0     Maharashtra   Western       Healthcare             Auto       1  23952.0  
1  Andhra Pradesh  Southern             Govt             Auto       3  23934.0  
2   Uttar Pradesh   Central       Automobile             Auto       3  23924.0  
3       Karnataka  Southern     Construction             Auto       2  23912.0  
4         Gujarat   Western  Food Processing             Auto       2  23877.0

### After Data Cleaning – Dataset Overview

After exploring and cleaning the Diwali Sales dataset, we gained the following insights about its structure and quality:

- The dataset contains **11,243 unique records** and **13 relevant features** after cleaning.
- There are **no missing values** in any column, making the dataset reliable for analysis.
- The data includes both:
  - **Numerical variables:** Age, Orders, Amount  
  - **Categorical variables:** Gender, Age Group, State, Zone, Occupation, Product Category
- Each row represents a **customer purchase transaction during the Diwali festive season**.
- Customer demographic information such as **age, gender, marital status, and occupation** is available.
- Sales-related details include **number of orders** and **total purchase amount**.
- Regional information like **state and zone** helps analyze geographical sales patterns.
- Overall, the dataset is **clean, consistent, and analysis-ready**, making it suitable for **exploratory data analysis and business insight generation**.


## 3. Understanding Your Variables

**What it is:**  
Examining each column to understand the type of data it holds (numerical or categorical) and what it represents in the context of the project.

**Why we use it:**  
To ensure we **know what each variable means**, decide which type of analysis or visualization is appropriate, and prepare for meaningful insights.


In [19]:
# Dataset Columns
df.columns

Index(['User_ID', 'Cust_name', 'Product_ID', 'Gender', 'Age_Group', 'Age',
       'Marital_Status', 'State', 'Zone', 'Occupation', 'Product_Category',
       'Orders', 'Amount'],
      dtype='object')

In [20]:
# Dataset Describe
df.describe(include='all')

User_ID Cust_name Product_ID Gender Age_Group           Age  \
count   1.124300e+04     11243      11243  11243     11243  11243.000000   
unique           NaN      1250       2351      2         7           NaN   
top              NaN  Vishakha  P00265242      F     26-35           NaN   
freq             NaN        42         53   7838      4538           NaN   
mean    1.003005e+06       NaN        NaN    NaN       NaN     35.422841   
std     1.716141e+03       NaN        NaN    NaN       NaN     12.756369   
min     1.000001e+06       NaN        NaN    NaN       NaN     12.000000   
25%     1.001494e+06       NaN        NaN    NaN       NaN     27.000000   
50%     1.003065e+06       NaN        NaN    NaN       NaN     33.000000   
75%     1.004430e+06       NaN        NaN    NaN       NaN     43.000000   
max     1.006040e+06       NaN        NaN    NaN       NaN     92.000000   

        Marital_Status          State     Zone Occupation    Product_Category  \
count     11243.000000          11243    11243      11243               11243   
unique             NaN             16        5         15                  18   
top                NaN  Uttar Pradesh  Central  IT Sector  Clothing & Apparel   
freq               NaN           1944     4294       1586                2653   
mean          0.420261            NaN      NaN        NaN                 NaN   
std           0.493623            NaN      NaN        NaN                 NaN   
min           0.000000            NaN      NaN        NaN                 NaN   
25%           0.000000            NaN      NaN        NaN                 NaN   
50%           0.000000            NaN      NaN        NaN                 NaN   
75%           1.000000            NaN      NaN        NaN                 NaN   
max           1.000000            NaN      NaN        NaN                 NaN   

              Orders        Amount  
count   11243.000000  11243.000000  
unique           NaN           NaN  
top              NaN           NaN  
freq             NaN           NaN  
mean        2.488749   9452.649331  
std         1.114960   5219.125950  
min         1.000000    188.000000  
25%         1.000000   5444.000000  
50%         2.000000   8109.000000  
75%         3.000000  12672.000000  
max         4.000000  23952.000000

### Column-wise Understanding

- **User_ID:** Unique identifier for each customer.  
  *Why:* Helps track individual transactions and customer-level analysis.

- **Cust_name:** Name of the customer.  
  *Why:* Mainly for reference; not used in numerical analysis.

- **Product_ID:** Unique identifier for each product.  
  *Why:* Useful for product-level analysis and revenue tracking.

- **Gender:** Customer's gender (M/F).  
  *Why:* Enables demographic analysis and gender-based purchasing trends.

- **Age_Group:** Customer's age range (e.g., 26-35).  
  *Why:* Helps understand purchasing behavior by age segment.

- **Age:** Exact age of the customer.  
  *Why:* Allows more precise demographic analysis and age-based segmentation.

- **Marital_Status:** Customer's marital status (0 = Single, 1 = Married).  
  *Why:* Useful to analyze buying patterns based on marital status.

- **State:** State where the customer resides.  
  *Why:* Supports regional sales analysis and location-based insights.

- **Zone:** Geographical zone (e.g., Northern, Southern).  
  *Why:* Helps group states into zones for broader regional analysis.

- **Occupation:** Customer's occupation (e.g., Healthcare, Govt).  
  *Why:* Allows analysis of purchasing trends by profession.

- **Product_Category:** Category of the purchased product (e.g., Electronics, Auto).  
  *Why:* Identifies high-performing product categories and sales trends.

- **Orders:** Number of items ordered per transaction.  
  *Why:* Helps calculate total quantity sold and understand order behavior.

- **Amount:** Total purchase amount for the transaction.  
  *Why:* Key metric for revenue analysis and financial insights.


## 4. Data Wrangling